src/DataStructures/Mesh.hpp

```diff
@@ -0,0 +1,176 @@
+ * a simulated physical field can be represented by its value on each grid
+ * point and then approximated by a polynomial of degree \f$p=N-1\f$ through a
+ * linear combination of Lagrange polynomials.
+ *
+ * \note A field represented by a `DataVector` has no meaning without an
+ * accompanying `Mesh` to provide context. Only w.r.t a `Mesh` can a
+ * `DataVector` of length `mesh.number_of_grid_points()` be interpreted as the
+ * field values at the collocation points, that are in turn identical to the
+ * nodal expansion coefficients in Lagrange polynomials.
+ *
+ * \tparam Dim the number of dimensions of the computational grid.
+ */
+template <size_t Dim>
+class Mesh {
+ public:
+  Mesh() noexcept = default;
+
+  /*!
+   * \brief Construct a computational grid with the same number of collocation
+   * points in each dimension.
+   *
+   * \param isotropic_extents The number of collocation points in each
+   * dimension.
+   * \param basis The choice of spectral basis to compute the
+   * collocation points
+   * \param quadrature The choice of quadrature to compute
+   * the collocation points
+   */
+  Mesh(const size_t isotropic_extents, const Spectral::Basis basis,
+       const Spectral::Quadrature quadrature) noexcept
+      : extents_(isotropic_extents) {
+    bases_.fill(basis);
+    quadratures_.fill(quadrature);
+  }
+
+  /*!
+   * Construct a computational grid where each dimension can have a different
+   * number of collocation points.
+   *
+   * \param extents The number of collocation points per dimension
+   * \param basis The choice of spectral basis to compute the
+   * collocation points
+   * \param quadrature The choice of quadrature to compute
+   * the collocation points
+   */
+  Mesh(std::array<size_t, Dim> extents, const Spectral::Basis basis,
+       const Spectral::Quadrature quadrature) noexcept
+      : extents_(std::move(extents)) {
+    bases_.fill(basis);
+    quadratures_.fill(quadrature);
+  }
+
+  /*!
+   * Construct a computational grid where each dimension can have both a
+   * different number and placement of collocation points.
+   *
+   * \param extents The number of collocation points per dimension
+   * \param bases The choice of spectral bases to compute the
+   * collocation points per dimension
+   * \param quadratures The choice of quadratures to compute
+   * the collocation points per dimension
+   */
+  Mesh(std::array<size_t, Dim> extents, std::array<Spectral::Basis, Dim> bases,
+       std::array<Spectral::Quadrature, Dim> quadratures) noexcept
+      : extents_(std::move(extents)),
+        bases_(std::move(bases)),
+        quadratures_(std::move(quadratures)) {}
+
+  /*!
+   * \brief The number of grid points in each dimension of the grid.
+   */
+  const Index<Dim>& extents() const noexcept { return extents_; }
+
+  /*!
+   * \brief The number of grid points in dimension `d` of the grid.
+   */
+  size_t extents(const size_t d) const noexcept { return extents_[d]; }
```

Comment:
Below you have `bases()` to return the array, `basis(d)` to return a single dimension. Same with `quadratures()` vs. `quadrature(d)`. For consistency, can this be singular `extent(d)` ?

Reply:
I think they should just be `basis()`, `basis(d)`, `quadrature()`, `quadrature(d)`, `extents()`, and `extents(d)`

Reply:
Having the plural return the array and the singular return the element as @fmahebert suggested makes most sense I think, don't you agree @nilsdeppe?

Reply:
I think we still say extents even in one dimension, e.g. "The extents in the x direction".

Reply:
This the English native speakers should determine, then tell me what to name these functions please ;)

Reply:
lets stick with basis, quadrature, and extents, overloading these for array access and individual element access